gemd/builders/impl.py

```diff
@@ -0,0 +1,424 @@
+            mass_fraction = NominalReal(nominal=mass_fraction, units='')
+        my_ingredient_run.mass_fraction = mass_fraction
+
+    if number_fraction is not None:
+        if isinstance(number_fraction, float):
+            number_fraction = NominalReal(nominal=number_fraction, units='')
+        my_ingredient_run.number_fraction = number_fraction
+
+    if volume_fraction is not None:
+        if isinstance(volume_fraction, float):
+            volume_fraction = NominalReal(nominal=volume_fraction, units='')
+        my_ingredient_run.volume_fraction = volume_fraction
+
+    if absolute_quantity is not None:
+        if isinstance(absolute_quantity, float):
+            if absolute_units is None:
+                raise ValueError("Absolute Units are required if Absolute Quantity is not a Value")
+            absolute_quantity = NominalReal(nominal=absolute_quantity, units=absolute_units)
+            absolute_units = None
+        my_ingredient_run.absolute_quantity = absolute_quantity
+
+    if absolute_units is not None:
+        raise ValueError("Absolute Units are only used if "
+                         "Absolute Quantity is given as is a float.")
+
+    return my_ingredient_run
+
+
+def add_measurement(material: MaterialRun,
+                    *,
+                    name: str = None,
+                    template: MeasurementTemplate = None,
+                    attributes: List[BaseAttribute] = None,
+                    ) -> MeasurementRun:
+    """
+    Add a measurement run-spec set to a MaterialRun.
+
+    Parameters
+    ----------
+    material: MaterialRun
+        The `material` for the returned MeasurementRun
+    name: str
+        The name of the measurement.  Defaults to
+        `template.name` if `template` is defined.
+    template: MeasurementTemplate
+        The MeasurementTemplate.
+    attributes: List[BaseAttribute]
+        Attributes you want associated with this MeasurementRun.
+
+    Returns
+    --------
+    MeasurementRun
+        A MeasurementRun with linked material, spec and template
+
+    """
+    if name is None:
+        if template is None:
+            raise ValueError("Either a name or a template must be provided")
+        else:
+            name = template.name
+
+    my_measurement_spec = MeasurementSpec(name, template=template)
+    my_measurement_run = MeasurementRun(name, spec=my_measurement_spec, material=material)
+
+    if attributes is not None:
+        for attribute in attributes:
+            if isinstance(attribute, Property):
+                my_measurement_run.properties.append(attribute)
+            elif isinstance(attribute, Condition):
+                my_measurement_run.conditions.append(attribute)
+            elif isinstance(attribute, Parameter):
+                my_measurement_run.parameters.append(attribute)
+            else:
+                raise ValueError(f"Unhandled Attribute type {type(attribute)}")
+
+    return my_measurement_run
+
+
+@overload
```

Comment:
The heavy use of `@overload` seems unnecessary. To take `add_attribute` as an example, the multiple overloaded methods enforce "A `PropertyTemplate` goes with a `HasProperties`, ..." but the implementation still has to check that the template is right for the target, so I don't see what is gained.

Reply:
Yeah, I was trying out a new toy.  I agree that it didn't return the quality of results I was hoping for.  At least for IntelliJ, the integration with the automated type checker didn't do quite as much I was hoping it would.  The better way to do this might be Generics, but even that probably doesn't buy us enough to justify the all the code.